ananse/network.py

```diff
@@ -623,178 +299,119 @@ def create_promoter_network(self, ddf, prom, fin_expression, outfile, column="tp
-            tfdf, filter_bed = self.clear_peak(ddf)
-            
-            prom = self.get_promoter_dataframe(filter_bed)
+        tfs.rename(
+            columns={"target": "tf", "target_expression": "tf_expression"}, inplace=True
+        )
 
-            logger.info("Create network")
-            self.create_promoter_network(ddf, prom, fin_expression, outfile)
+        expression["key"] = 0
+        tfs["key"] = 0
 
-        else:
-            logger.info("Read data")
-            ddf = dd.read_csv(binding, sep="\t")[["factor", "enhancer", "binding"]]
+        network = expression.merge(tfs, how="outer")
+        network = network[["tf", "target", "tf_expression", "target_expression"]]
 
-            tfdf, filter_bed = self.clear_peak(ddf)
+        for col in ["tf_expression", "target_expression"]:
+            if rank:
+                network[col] = rankdata(network[col])
+            network[col] = minmax_scale(network[col])
 
-            prom = self.get_promoter_dataframe(filter_bed)
-            p = self.get_gene_dataframe(filter_bed)
+        network["tf_target"] = network["tf"] + "_" + network["target"]
+        network = network.set_index("tf_target").drop(columns=["tf", "target"])
+        logger.info("creating expression dataframe")
+        network = dd.from_pandas(network, npartitions=30)
+        # print(network)
+        return network
 
-            weight = self.distance_weight(self.include_promoter, self.include_enhancer)
+    def unique_enhancers(self, binding_fname, chrom=None):
+        p = re.compile("[:-]")
+        logger.info("reading enhancers")
+        enhancers = pd.read_table(binding_fname, usecols=["enhancer"])["enhancer"]
+        # enhancers = pd.read_table("/ceph/rimlsfnwi/data/moldevbio/heeringen/heeringen/atac_ananse/models/saved/H3K27ac.qnorm.ref.txt", usecols=[0]).iloc[:,0]
+
+        if chrom:
+            logger.info(f"filtering for {chrom}")
+            enhancers = enhancers[enhancers.str.contains(f"^{chrom}:")]
+
+        enhancers = enhancers.unique()
+        logger.info("writing temporary enhancers")
+        f = NamedTemporaryFile(mode="w+", dir=mytmpdir(), delete=False)
+        for e in enhancers:
+            print("{}\t{}\t{}".format(*re.split(p, e)), file=f)
+
+        # print(open(f.name).readlines()[:10])
+
+        return f.name
+
+    def run_network(
+        self,
+        binding,
+        fin_expression=None,
+        tfs=None,
+        corrfiles=None,
+        outfile=None,
+        up=1e5,
+        down=1e5,
+        alpha=None,
+        promoter=2000,
+        full_weight_region=5000,
+    ):
+        # Expression base network
+        logger.info("Loading expression")
+        df_expression = self.create_expression_network(
+            fin_expression, tfs=tfs, rank=True
+        )
 
+        # Use a version of the binding network, either promoter-based, enhancer-based
+        # or both.
+        if self.include_promoter or self.include_enhancer:
```

Comment:
this if else structure contains the following three options:
- if only promoters/enhancers are used
- if `fun_expression` is None.
- if both are used

the first two are currently merged. Perhaps it is clearer if they are separated? That way we can also skip `create_expression_network` if the input is `None` anyway

Reply:
> A couple of remarks, not all related to this PR, but more to the network.py file in general. The PR is nice
> 
> * There is a lot of stuff being written and read to disk, which often seems unnecessary. Perhaps switching from pybedtools to pyranges might solve some of these issues.

Yes, we should see where this is not necessary. One thing I would like to keep in mind is the memory tradeoff.. it should also be usable on a machine with reasonable RAM (say 16-32G)

> * I never worked with dask, not sure how some stuff works. Are some computations "futures"? Which you do later?

Yes! It is now written that most the computation is actually performed on the `compute()` call. This means we can later manage this by setting up a dask.distributed client that manages workers and memory consumption.

> * Variable and function names are often unclear. This, in combination with the lack of comments and docstrings, make the code quite hard to read sometimes, even though the code is really neat. The stuff that happens in network isn't overly complicated, but every time I look at a function I have no clue what it happens until I read it carefully. Adding docstrings to functions with an example input and output would make reading the code much easier.

I know.. I've started adding comments and updating variable names, but there's still a lot to do there!